libs/vkd3d/cache.c

```diff
@@ -35,75 +165,697 @@ VkResult vkd3d_create_pipeline_cache(struct d3d12_device *device,
-        return hresult_from_vk_result(vr);
+        aligned_chunk_size = align(chunk->size + sizeof(struct vkd3d_pipeline_blob_chunk),
+                VKD3D_PIPELINE_BLOB_CHUNK_ALIGN);
+        if (aligned_chunk_size > size)
+            return NULL;
+        if (chunk->type == type)
+            return chunk;
+
+        chunk = (const struct vkd3d_pipeline_blob_chunk *)&chunk->data[align(chunk->size, VKD3D_PIPELINE_BLOB_CHUNK_ALIGN)];
+        size -= aligned_chunk_size;
     }
 
+    return NULL;
+}
+
+HRESULT d3d12_cached_pipeline_state_validate(struct d3d12_device *device,
+        const struct d3d12_cached_pipeline_state *state,
+        const struct vkd3d_pipeline_cache_compatibility *compat)
+{
+    const VkPhysicalDeviceProperties *device_properties = &device->device_info.properties2.properties;
+    const struct vkd3d_pipeline_blob *blob = state->blob.pCachedBlob;
+    const struct vkd3d_pipeline_blob_chunk_pso_compat *pso_compat;
+    const struct vkd3d_pipeline_blob_chunk *chunk;
+    size_t payload_size;
+    uint32_t checksum;
+
     /* Avoid E_INVALIDARG with an invalid header size, since that may confuse some games */
     if (state->blob.CachedBlobSizeInBytes < sizeof(*blob) || blob->version != VKD3D_CACHE_BLOB_VERSION)
         return D3D12_ERROR_DRIVER_VERSION_MISMATCH;
 
+    payload_size = state->blob.CachedBlobSizeInBytes - offsetof(struct vkd3d_pipeline_blob, data);
+
     /* Indicate that the cached data is not useful if we're running on a different device or driver */
     if (blob->vendor_id != device_properties->vendorID || blob->device_id != device_properties->deviceID)
         return D3D12_ERROR_ADAPTER_NOT_FOUND;
 
-    /* Check the vkd3d build since the shader compiler itself may change,
+    /* Check the vkd3d-proton build since the shader compiler itself may change,
      * and the driver since that will affect the generated pipeline cache.
      * Based on global configuration flags, which extensions are available, etc,
      * the generated shaders may also change, so key on that as well. */
     if (blob->vkd3d_build != vkd3d_build ||
             blob->vkd3d_shader_interface_key != device->shader_interface_key ||
-            memcmp(blob->cache_uuid, device_properties->pipelineCacheUUID, VK_UUID_SIZE))
+            memcmp(blob->cache_uuid, device_properties->pipelineCacheUUID, VK_UUID_SIZE) != 0)
+        return D3D12_ERROR_DRIVER_VERSION_MISMATCH;
+
+    checksum = vkd3d_pipeline_blob_compute_data_checksum(blob->data, payload_size);
+
+    if (checksum != blob->checksum)
+    {
+        ERR("Corrupt PSO cache blob entry found!\n");
+        /* Same rationale as above, avoid E_INVALIDARG, since that may confuse some games */
         return D3D12_ERROR_DRIVER_VERSION_MISMATCH;
+    }
+
+    /* Fetch compat info. */
+    chunk = find_blob_chunk(CONST_CAST_CHUNK_BASE(blob), payload_size, VKD3D_PIPELINE_BLOB_CHUNK_TYPE_PSO_COMPAT);
+    if (!chunk || chunk->size != sizeof(*pso_compat))
+        return E_FAIL;
```

Comment:
I'm not entirely sure about the return codes here (and below), what's the rationale for those?

Reply:
The E_INVALIDARGs are supposed to catch the error in D3D12 API. If you pass mismatching descriptions compared to the blob, API requires us to fail the call.

The E_FAIL should be seen more as an assertion. I think I made it that error code since there was not other logical error code to use, but I can make that D3D12_ERROR_DRIVER_VERSION_MISMATCH I think.

Reply:
The validate call is made early in the Create*Pipeline calls, and it's supposed to catch any invalid usage early. We cannot keep trucking on and build SPIR-V files late.

Reply:
Fixed.